File: scripts/sync-google-ads.mjs
#!/usr/bin/env node
/**
 * Google Ads Sync Script
 * Fetches Google Ads data from MCC account and populates ClientGoogleAdsSnapshot
 * 
 * Prerequisites:
 *   - GOOGLE_ADS_DEVELOPER_TOKEN in .env.local
 *   - GOOGLE_ADS_REFRESH_TOKEN in .env.local
 *   - GOOGLE_ADS_CLIENT_ID in .env.local
 *   - GOOGLE_ADS_CLIENT_SECRET in .env.local
 *   - GOOGLE_ADS_MCC_CUSTOMER_ID in .env.local (optional, if using MCC)
 * 
 * Usage:
 *   node scripts/sync-google-ads.mjs [--month YYYY-MM] [--customer-id 1234567890]
 */

import { PrismaClient } from '@prisma/client';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Load environment variables
dotenv.config({ path: join(__dirname, '..', '.env.local') });

const prisma = new PrismaClient();

// ============================================================================
// CONFIGURATION
// ============================================================================

const GOOGLE_ADS_DEVELOPER_TOKEN = process.env.GOOGLE_ADS_DEVELOPER_TOKEN;
const GOOGLE_ADS_REFRESH_TOKEN = process.env.GOOGLE_ADS_REFRESH_TOKEN;
const GOOGLE_ADS_CLIENT_ID = process.env.GOOGLE_ADS_CLIENT_ID;
const GOOGLE_ADS_CLIENT_SECRET = process.env.GOOGLE_ADS_CLIENT_SECRET;
const GOOGLE_ADS_MCC_CUSTOMER_ID = process.env.GOOGLE_ADS_MCC_CUSTOMER_ID;

// ============================================================================
// GOOGLE ADS API CLIENT PLACEHOLDER
// ============================================================================

/**
 * Fetch Google Ads data for a customer account
 * @param {string} customerId - Google Ads customer ID (no dashes)
 * @param {object} dateRange - { start: 'YYYY-MM-DD', end: 'YYYY-MM-DD' }
 * @returns {Promise<object>} Ads data object
 * 
 * TODO: Implement actual Google Ads API call using google-ads-api npm package
 * This is a placeholder that returns null until OAuth token is available
 */
async function fetchAdsData(customerId, dateRange) {
  // Verify credentials exist
  if (!GOOGLE_ADS_DEVELOPER_TOKEN || !GOOGLE_ADS_REFRESH_TOKEN) {
    console.warn(`⚠️  Missing Google Ads credentials for customer ${customerId}`);
    return null;
  }

  // TODO: Implement Google Ads API call
  // Example implementation:
  /*
  const { GoogleAdsApi } = await import('google-ads-api');
  
  const client = new GoogleAdsApi({
    client_id: GOOGLE_ADS_CLIENT_ID,
    client_secret: GOOGLE_ADS_CLIENT_SECRET,
    developer_token: GOOGLE_ADS_DEVELOPER_TOKEN,
  });

  const customer = client.Customer({
    customer_id: customerId,
    refresh_token: GOOGLE_ADS_REFRESH_TOKEN,
  });

  const query = `
    SELECT
      campaign.id,
      campaign.name,
      campaign.status,
      metrics.impressions,
      metrics.clicks,
      metrics.cost_micros,
      metrics.conversions,
      metrics.conversions_value,
      metrics.search_impression_share,
      segments.date
    FROM campaign
    WHERE segments.date BETWEEN '${dateRange.start}' AND '${dateRange.end}'
      AND campaign.status = 'ENABLED'
  `;

  const campaigns = await customer.query(query);
  
  // Aggregate metrics
  let totalImpressions = 0;
  let totalClicks = 0;
  let totalSpendMicros = 0;
  let totalConversions = 0;
  let impressionShareSum = 0;
  let impressionShareCount = 0;
  const activeCampaignNames = new Set();
  const keywords = new Set(); // Would need separate query for keywords

  for (const row of campaigns) {
    totalImpressions += row.metrics.impressions || 0;
    totalClicks += row.metrics.clicks || 0;
    totalSpendMicros += row.metrics.cost_micros || 0;
    totalConversions += row.metrics.conversions || 0;
    
    if (row.metrics.search_impression_share) {
      impressionShareSum += row.metrics.search_impression_share;
      impressionShareCount++;
    }
    
    if (row.campaign.status === 'ENABLED') {
      activeCampaignNames.add(row.campaign.name);
    }
  }

  const totalSpend = totalSpendMicros / 1000000; // Convert micros to dollars
  const costPerConversion = totalConversions > 0 ? totalSpend / totalConversions : 0;
  const conversionRate = totalClicks > 0 ? totalConversions / totalClicks : 0;
  const avgImpressionShare = impressionShareCount > 0 ? impressionShareSum / impressionShareCount : 0;

  return {
    impressions: totalImpressions,
    clicks: totalClicks,
    spend: totalSpend,
    conversions: totalConversions,
    costPerConversion,
    conversionRate,
    impressionShare: avgImpressionShare,
    activeCampaigns: activeCampaignNames.size,
    campaignNames: Array.from(activeCampaignNames),
    topKeywords: [], // Would populate from separate query
    budgetUtilization: 0 // Would calculate if budget data available
  };
  */

  console.log(`📊 [PLACEHOLDER] Would fetch ads data for customer ${customerId} (${dateRange.start} to ${dateRange.end})`);
  return null;
}

/**
 * Fetch list of all customer accounts under MCC
 * @returns {Promise<string[]>} Array of customer IDs
 * 
 * TODO: Implement MCC customer list fetch
 */
async function fetchMCCCustomerList() {
  if (!GOOGLE_ADS_MCC_CUSTOMER_ID) {
    console.warn('⚠️  No MCC customer ID configured, skipping MCC fetch');
    return [];
  }

  // TODO: Implement MCC account fetch
  /*
  const { GoogleAdsApi } = await import('google-ads-api');
  
  const client = new GoogleAdsApi({
    client_id: GOOGLE_ADS_CLIENT_ID,
    client_secret: GOOGLE_ADS_CLIENT_SECRET,
    developer_token: GOOGLE_ADS_DEVELOPER_TOKEN,
  });

  const mcc = client.Customer({
    customer_id: GOOGLE_ADS_MCC_CUSTOMER_ID,
    refresh_token: GOOGLE_ADS_REFRESH_TOKEN,
  });

  const query = `
    SELECT
      customer_client.id,
      customer_client.descriptive_name,
      customer_client.manager
    FROM customer_client
    WHERE customer_client.status = 'ENABLED'
      AND customer_client.manager = FALSE
  `;

  const customers = await mcc.query(query);
  return customers.map(c => c.customer_client.id.toString());
  */

  console.log('📋 [PLACEHOLDER] Would fetch MCC customer list');
  return [];
}

// ============================================================================
// DATABASE OPERATIONS
// ============================================================================

/**
 * Map Google Ads customer ID to ClientProfile by acronym
 * @param {string} customerId - Google Ads customer ID
 * @returns {Promise<object|null>} ClientProfile or null
 */
async function findClientByCustomerId(customerId) {
  // First try exact match on googleAdsCustomerId
  let client = await prisma.clientProfile.findFirst({
    where: {
      googleAdsCustomerId: customerId,
      tenantId: 'gyc'
    }
  });

  if (client) return client;

  // TODO: Implement fuzzy matching logic
  // Could match by company name from Google Ads account name
  
  return null;
}

/**
 * Upsert Google Ads snapshot for a client
 */
async function upsertAdsSnapshot(data) {
  const {
    clientId,
    companyAcronym,
    customerId,
    periodMonth,
    adsData
  } = data;

  if (!adsData) {
    console.log(`  ⏭️  Skipping ${companyAcronym} - no ads data available`);
    return null;
  }

  try {
    const snapshot = await prisma.clientGoogleAdsSnapshot.upsert({
      where: {
        tenantId_companyAcronym_periodMonth: {
          tenantId: 'gyc',
          companyAcronym,
          periodMonth
        }
      },
      update: {
        customerId,
        impressions: adsData.impressions,
        clicks: adsData.clicks,
        spend: adsData.spend,
        conversions: adsData.conversions,
        costPerConversion: adsData.costPerConversion,
        conversionRate: adsData.conversionRate,
        impressionShare: adsData.impressionShare,
        activeCampaigns: adsData.activeCampaigns,
        campaignNames: adsData.campaignNames,
        topKeywords: adsData.topKeywords || [],
        budgetUtilization: adsData.budgetUtilization || 0,
        syncedAt: new Date(),
        dataSource: 'google-ads-api'
      },
      create: {
        tenantId: 'gyc',
        clientId,
        companyAcronym,
        customerId,
        periodMonth,
        impressions: adsData.impressions,
        clicks: adsData.clicks,
        spend: adsData.spend,
        conversions: adsData.conversions,
        costPerConversion: adsData.costPerConversion,
        conversionRate: adsData.conversionRate,
        impressionShare: adsData.impressionShare,
        activeCampaigns: adsData.activeCampaigns,
        campaignNames: adsData.campaignNames,
        topKeywords: adsData.topKeywords || [],
        budgetUtilization: adsData.budgetUtilization || 0,
        dataSource: 'google-ads-api'
      }
    });

    console.log(`  ✅ Upserted ${companyAcronym} for ${periodMonth}`);
    return snapshot;
  } catch (error) {
    console.error(`  ❌ Failed to upsert ${companyAcronym}:`, error.message);
    return null;
  }
}

// ============================================================================
// SYNC LOGIC
// ============================================================================

async function syncGoogleAds(options = {}) {
  const {
    month = new Date().toISOString().slice(0, 7), // Default to current month
    customerId = null // Optional: sync specific customer only
  } = options;

  console.log('🚀 Starting Google Ads sync...');
  console.log(`   Month: ${month}`);
  if (customerId) console.log(`   Customer ID: ${customerId}`);

  // Verify credentials
  if (!GOOGLE_ADS_DEVELOPER_TOKEN || !GOOGLE_ADS_REFRESH_TOKEN) {
    console.error('❌ Missing Google Ads credentials. Set environment variables:');
    console.error('   - GOOGLE_ADS_DEVELOPER_TOKEN');
    console.error('   - GOOGLE_ADS_REFRESH_TOKEN');
    console.error('   - GOOGLE_ADS_CLIENT_ID');
    console.error('   - GOOGLE_ADS_CLIENT_SECRET');
    process.exit(1);
  }

  // Calculate date range for the month
  const [year, monthNum] = month.split('-');
  const startDate = `${year}-${monthNum}-01`;
  const lastDay = new Date(year, monthNum, 0).getDate();
  const endDate = `${year}-${monthNum}-${lastDay}`;
  const dateRange = { start: startDate, end: endDate };

  console.log(`   Date range: ${startDate} to ${endDate}`);

  let customerIds = [];

  if (customerId) {
    // Sync specific customer
    customerIds = [customerId];
  } else {
    // Fetch all customers from MCC or database
    console.log('\n📋 Fetching customer list...');
    
    // Try MCC first
    const mccCustomers = await fetchMCCCustomerList();
    
    if (mccCustomers.length > 0) {
      customerIds = mccCustomers;
      console.log(`   Found ${customerIds.length} customers in MCC`);
    } else {
      // Fallback: use customers already in database
      const clients = await prisma.clientProfile.findMany({
        where: {
          googleAdsCustomerId: { not: null },
          tenantId: 'gyc'
        },
        select: {
          googleAdsCustomerId: true
        }
      });
      customerIds = clients.map(c => c.googleAdsCustomerId).filter(Boolean);
      console.log(`   Found ${customerIds.length} customers in database`);
    }
  }

  if (customerIds.length === 0) {
    console.log('\n⚠️  No customers to sync');
    return { synced: 0, failed: 0 };
  }

  // Sync each customer
  console.log(`\n📊 Syncing ${customerIds.length} customer(s)...`);
  
  let synced = 0;
  let failed = 0;

  for (const custId of customerIds) {
    console.log(`\n🔄 Processing customer ${custId}...`);

    // Find matching ClientProfile
    const client = await findClientByCustomerId(custId);
    
    if (!client) {
      console.log(`  ⚠️  No matching ClientProfile found for customer ${custId}`);
      failed++;
      continue;
    }

    console.log(`  📌 Matched to: ${client.acronym} (${client.companyName})`);

    // Fetch ads data
    const adsData = await fetchAdsData(custId, dateRange);

    // Upsert snapshot
    const result = await upsertAdsSnapshot({
      clientId: client.id,
      companyAcronym: client.acronym,
      customerId: custId,
      periodMonth: month,
      adsData
    });

    if (result) {
      synced++;
    } else {
      failed++;
    }
  }

  console.log('\n✅ Sync complete!');
  console.log(`   Synced: ${synced}`);
  console.log(`   Failed: ${failed}`);

  return { synced, failed };
}

// ============================================================================
// CLI
// ============================================================================

async function main() {
  const args = process.argv.slice(2);
  const options = {};

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--month' && args[i + 1]) {
      options.month = args[i + 1];
      i++;
    } else if (args[i] === '--customer-id' && args[i + 1]) {
      options.customerId = args[i + 1];
      i++;
    }
  }

  try {
    await syncGoogleAds(options);
  } catch (error) {
    console.error('\n❌ Sync failed:', error);
    throw error;
  } finally {
    await prisma.$disconnect();
  }
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main()
    .then(() => {
      console.log('\n👋 Done');
      process.exit(0);
    })
    .catch(err => {
      console.error('\n💥 Fatal error:', err);
      process.exit(1);
    });
}

export { syncGoogleAds, fetchAdsData, findClientByCustomerId };

File: scripts/sync-google-ads.py
#!/usr/bin/env python3
"""
Google Ads Sync Script
Pulls last 30 days vs previous 30 days for all ENABLED accounts under MCC 4782502607
Upserts results into the GoogleAdsAccount table via POST to /api/google-ads/sync
Resume-safe — picks up where it left off
Designed to run nightly via cron
"""

import os
import sys
import yaml
import requests
from datetime import datetime, timedelta
from google.ads.googleads.client import GoogleAdsClient
from google.ads.googleads.errors import GoogleAdsException

# Configuration
GOOGLE_ADS_YAML = os.path.join(os.path.dirname(__file__), '..', 'google-ads.yaml')
API_ENDPOINT = 'http://localhost:3000/api/google-ads/sync'
MCC_CUSTOMER_ID = '4782502607'

def load_client():
    """Load Google Ads client from google-ads.yaml"""
    if not os.path.exists(GOOGLE_ADS_YAML):
        print(f'❌ Google Ads config not found: {GOOGLE_ADS_YAML}')
        sys.exit(1)
    
    return GoogleAdsClient.load_from_storage(GOOGLE_ADS_YAML)

def get_enabled_accounts(client, mcc_id):
    """Get all enabled customer accounts under the MCC"""
    ga_service = client.get_service("GoogleAdsService")
    
    query = """
        SELECT
          customer_client.id,
          customer_client.descriptive_name,
          customer_client.status
        FROM customer_client
        WHERE customer_client.status = 'ENABLED'
    """
    
    try:
        response = ga_service.search(customer_id=mcc_id, query=query)
        accounts = []
        for row in response:
            accounts.append({
                'id': str(row.customer_client.id),
                'name': row.customer_client.descriptive_name
            })
        return accounts
    except GoogleAdsException as ex:
        print(f'❌ Error fetching accounts: {ex}')
        return []

def get_account_metrics(client, customer_id, start_date, end_date):
    """Get metrics for a single account for the given date range"""
    ga_service = client.get_service("GoogleAdsService")
    
    query = f"""
        SELECT
          metrics.cost_micros,
          metrics.clicks,
          metrics.impressions,
          metrics.ctr
        FROM campaign
        WHERE segments.date BETWEEN '{start_date}' AND '{end_date}'
    """
    
    try:
        response = ga_service.search(customer_id=customer_id, query=query)
        
        total_cost = 0
        total_clicks = 0
        total_impressions = 0
        
        for row in response:
            total_cost += row.metrics.cost_micros / 1_000_000  # Convert micros to dollars
            total_clicks += row.metrics.clicks
            total_impressions += row.metrics.impressions
        
        cpc = total_cost / total_clicks if total_clicks > 0 else 0
        ctr = (total_clicks / total_impressions * 100) if total_impressions > 0 else 0
        
        return {
            'cost': total_cost,
            'clicks': total_clicks,
            'impressions': total_impressions,
            'cpc': cpc,
            'ctr': ctr
        }
    except GoogleAdsException as ex:
        print(f'  ⚠️  Error fetching metrics for {customer_id}: {ex}')
        return None

def calculate_change(curr, prev):
    """Calculate percentage change between current and previous"""
    if prev == 0:
        return None
    return ((curr - prev) / prev) * 100

def sync_account_to_api(account_data):
    """POST account data to the API sync endpoint"""
    try:
        response = requests.post(API_ENDPOINT, json={'accounts': [account_data]}, timeout=30)
        response.raise_for_status()
        return True
    except Exception as ex:
        print(f'  ❌ API sync error: {ex}')
        return False

def main():
    print('📊 Google Ads Sync Starting...\n')
    
    # Date ranges: last 30 days vs prior 30 days
    today = datetime.now().date()
    curr_end = today - timedelta(days=1)
    curr_start = curr_end - timedelta(days=29)
    prev_end = curr_start - timedelta(days=1)
    prev_start = prev_end - timedelta(days=29)
    
    print(f'Current period: {curr_start} to {curr_end}')
    print(f'Previous period: {prev_start} to {prev_end}\n')
    
    # Load client
    client = load_client()
    print(f'✓ Loaded Google Ads client\n')
    
    # Get all enabled accounts
    print(f'Fetching accounts under MCC {MCC_CUSTOMER_ID}...')
    accounts = get_enabled_accounts(client, MCC_CUSTOMER_ID)
    print(f'✓ Found {len(accounts)} enabled accounts\n')
    
    synced = 0
    errors = 0
    all_account_data = []
    
    for i, account in enumerate(accounts, 1):
        print(f'[{i}/{len(accounts)}] {account["name"]} ({account["id"]})')
        
        # Get current period metrics
        curr_metrics = get_account_metrics(client, account['id'], curr_start.strftime('%Y-%m-%d'), curr_end.strftime('%Y-%m-%d'))
        if not curr_metrics:
            errors += 1
            continue
        
        # Get previous period metrics
        prev_metrics = get_account_metrics(client, account['id'], prev_start.strftime('%Y-%m-%d'), prev_end.strftime('%Y-%m-%d'))
        if not prev_metrics:
            errors += 1
            continue
        
        # Calculate changes
        cpc_change = calculate_change(curr_metrics['cpc'], prev_metrics['cpc'])
        clicks_change = calculate_change(curr_metrics['clicks'], prev_metrics['clicks'])
        impressions_change = calculate_change(curr_metrics['impressions'], prev_metrics['impressions'])
        
        # Determine flags
        flags = []
        flagged = False
        
        if cpc_change and cpc_change > 100:
            flags.append(f'CPC up {int(cpc_change)}%')
            flagged = True
        elif cpc_change and cpc_change > 50:
            flags.append(f'CPC up {int(cpc_change)}%')
            flagged = True
        elif cpc_change and cpc_change > 20:
            flags.append(f'CPC up {int(cpc_change)}%')
            flagged = True
        
        if clicks_change and clicks_change < -30:
            flags.append(f'Clicks down {int(abs(clicks_change))}%')
            flagged = True
        
        if impressions_change and impressions_change < -30:
            flags.append(f'Impressions down {int(abs(impressions_change))}%')
            flagged = True
        
        # Prepare API payload
        account_data = {
            'accountId': account['id'],
            'accountName': account['name'],
            'currSpend': curr_metrics['cost'],
            'currClicks': curr_metrics['clicks'],
            'currImpressions': curr_metrics['impressions'],
            'currCpc': curr_metrics['cpc'],
            'currCtr': curr_metrics['ctr'],
            'prevSpend': prev_metrics['cost'],
            'prevClicks': prev_metrics['clicks'],
            'prevImpressions': prev_metrics['impressions'],
            'prevCpc': prev_metrics['cpc'],
            'cpcChange': cpc_change,
            'clicksChange': clicks_change,
            'impressionsChange': impressions_change,
            'flagged': flagged,
            'flags': flags
        }
        
        # Sync to API
        if sync_account_to_api(account_data):
            synced += 1
            all_account_data.append(account_data)
            print(f'  ✓ Synced')
        else:
            errors += 1
        
        print()
    
    # Update monthly snapshot for current month
    if all_account_data:
        print('\nUpdating monthly snapshot...')
        month_key = today.strftime('%Y-%m')
        month_label = today.strftime('%b %Y')
        
        total_spend = sum(a['currSpend'] for a in all_account_data)
        total_clicks = sum(a['currClicks'] for a in all_account_data)
        total_impressions = sum(a['currImpressions'] for a in all_account_data)
        active_count = len([a for a in all_account_data if a['currClicks'] > 0])
        avg_cpc = total_spend / total_clicks if total_clicks > 0 else 0
        
        monthly_data = {
            'monthKey': month_key,
            'monthLabel': month_label,
            'isPartial': True,
            'spend': total_spend,
            'clicks': total_clicks,
            'impressions': total_impressions,
            'avgCpc': avg_cpc,
            'accountCount': active_count,
        }
        
        try:
            dashboard_url = os.environ.get('DASHBOARD_URL', 'http://localhost:3000')
            response = requests.post(f'{dashboard_url}/api/google-ads/sync-monthly', json=monthly_data, timeout=30)
            response.raise_for_status()
            print(f'✓ Updated monthly snapshot for {month_label}')
        except Exception as ex:
            print(f'⚠️  Failed to update monthly snapshot: {ex}')
    
    print(f'\n✅ Sync complete!')
    print(f'   Synced: {synced}')
    print(f'   Errors: {errors}')

if __name__ == '__main__':
    main()
